File: packages/goblins/testing/src/node/postgres.ts
import type { Client } from "pg";
import {
	GenericContainer,
	type StartedTestContainer,
	Wait,
} from "testcontainers";

export interface PostgresOptions {
	/** Docker image to use (default: postgres:16-alpine) */
	image?: string;
	/** Database name (default: test_db) */
	database?: string;
	/** Username (default: test_user) */
	username?: string;
	/** Password (default: test_password) */
	password?: string;
	/** SQL files to run on startup */
	initScripts?: string[];
	/** Port to expose (0 = random) */
	port?: number;
}

export class PostgresTestcontainer {
	private container: StartedTestContainer;
	private options: Required<Omit<PostgresOptions, "initScripts" | "port">> & {
		initScripts?: string[];
		port?: number;
	};
	private _client?: Client;

	private constructor(
		container: StartedTestContainer,
		options: Required<Omit<PostgresOptions, "initScripts" | "port">> & {
			initScripts?: string[];
			port?: number;
		},
	) {
		this.container = container;
		this.options = options;
	}

	static async start(
		options: PostgresOptions = {},
	): Promise<PostgresTestcontainer> {
		const opts = {
			image: options.image ?? "postgres:16-alpine",
			database: options.database ?? "test_db",
			username: options.username ?? "test_user",
			password: options.password ?? "test_password",
			initScripts: options.initScripts,
			port: options.port,
		};

		let containerBuilder = new GenericContainer(opts.image)
			.withEnvironment({
				POSTGRES_DB: opts.database,
				POSTGRES_USER: opts.username,
				POSTGRES_PASSWORD: opts.password,
			})
			.withExposedPorts(opts.port ?? 5432)
			.withWaitStrategy(
				Wait.forLogMessage(/database system is ready to accept connections/),
			);

		// Add init scripts if provided
		if (opts.initScripts && opts.initScripts.length > 0) {
			for (const script of opts.initScripts) {
				containerBuilder = containerBuilder.withCopyFilesToContainer([
					{ source: script, target: `/docker-entrypoint-initdb.d/${script}` },
				]);
			}
		}

		const container = await containerBuilder.start();

		return new PostgresTestcontainer(container, opts);
	}

	async stop(): Promise<void> {
		if (this._client) {
			await this._client.end();
			this._client = undefined;
		}
		await this.container.stop();
	}

	getClient(): Client {
		if (!this._client) {
			// Lazy import to avoid dependency when not used
			const pg = require("pg");
			this._client = new pg.Client({
				host: this.getHost(),
				port: this.getPort(),
				database: this.options.database,
				user: this.options.username,
				password: this.options.password,
			});
			this._client.connect();
		}
		return this._client;
	}

	getConnectionString(): string {
		return `postgresql://${this.options.username}:${this.options.password}@${this.getHost()}:${this.getPort()}/${this.options.database}`;
	}

	getHost(): string {
		return this.container.getHost();
	}

	getPort(): number {
		return this.container.getMappedPort(this.options.port ?? 5432);
	}
}

File: packages/goblins/testing/src/python/postgres.py
"""PostgreSQL Testcontainer for Python integration tests."""

from typing import Optional
from testcontainers.postgres import PostgresContainer as TCPostgresContainer
import asyncpg


class PostgresContainer:
    """PostgreSQL container for integration testing with asyncpg."""

    def __init__(
        self,
        image: str = "postgres:16-alpine",
        username: str = "test_user",
        password: str = "test_password",
        database: str = "test_db",
        port: int = 5432
    ):
        self.image = image
        self.username = username
        self.password = password
        self.database = database
        self.port = port
        self._container: Optional[TCPostgresContainer] = None
        self._connection: Optional[asyncpg.Connection] = None

    async def start(self) -> "PostgresContainer":
        """Start the PostgreSQL container."""
        self._container = TCPostgresContainer(
            image=self.image,
            username=self.username,
            password=self.password,
            dbname=self.database,
            port=self.port
        )
        self._container.start()
        return self

    async def stop(self) -> None:
        """Stop the PostgreSQL container."""
        if self._connection:
            await self._connection.close()
            self._connection = None

        if self._container:
            self._container.stop()
            self._container = None

    async def get_connection(self) -> asyncpg.Connection:
        """Get an asyncpg connection to the database."""
        if not self._connection:
            if not self._container:
                raise RuntimeError("Container not started. Call start() first.")

            self._connection = await asyncpg.connect(
                host=self.get_host(),
                port=self.get_port(),
                user=self.username,
                password=self.password,
                database=self.database
            )

        return self._connection

    def get_connection_string(self) -> str:
        """Get the PostgreSQL connection string."""
        return f"postgresql://{self.username}:{self.password}@{self.get_host()}:{self.get_port()}/{self.database}"

    def get_host(self) -> str:
        """Get the container host."""
        if not self._container:
            raise RuntimeError("Container not started")
        return self._container.get_container_host_ip()

    def get_port(self) -> int:
        """Get the mapped port."""
        if not self._container:
            raise RuntimeError("Container not started")
        return int(self._container.get_exposed_port(self.port))

    async def __aenter__(self) -> "PostgresContainer":
        """Async context manager entry."""
        return await self.start()

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        """Async context manager exit."""
        await self.stop()
